include/experimental/__p2630_bits/submdspan_mapping.hpp

```diff
@@ -93,25 +81,71 @@ construct_sub_strides(const SrcMapping &src_mapping,
 namespace detail {
 
 // Figure out whether to preserve layout_left
-template <class IndexSequence, size_t SubRank, class... SliceSpecifiers>
-struct preserve_layout_left_mapping;
-
-template <class... SliceSpecifiers, size_t... Idx, size_t SubRank>
-struct preserve_layout_left_mapping<std::index_sequence<Idx...>, SubRank,
-                                    SliceSpecifiers...> {
-  constexpr static bool value =
-      // Preserve layout for rank 0
+template <class IndexType, size_t SubRank, class IndexSequence,
+          class... SliceSpecifiers>
+struct deduce_layout_left_submapping;
+
+template <class IndexType, size_t SubRank, size_t... Idx,
+          class... SliceSpecifiers>
+struct deduce_layout_left_submapping<
+    IndexType, SubRank, std::index_sequence<Idx...>, SliceSpecifiers...> {
+
+  using CountRange = index_sequence_scan_impl<
+      0, (std::is_convertible_v<SliceSpecifiers, IndexType> ? 0 : 1)...>;
+  //__static_partial_sums<!std::is_convertible_v<SliceSpecifiers,
+  // IndexType>...>;
+  constexpr static int NumGaps =
+      (((Idx > 0 && CountRange::get(Idx) == 1 &&
+         std::is_convertible_v<SliceSpecifiers, IndexType>)
+            ? 1
+            : 0) +
+       ... + 0);
+
+  constexpr static bool layout_left_value =
+      // Use layout_left for rank 0
+      (SubRank == 0) ||
+      // Use layout_left for rank 1 if leftmost slice specifier is range like
+      ((SubRank == 1) &&
+       ((Idx > 0 || std::is_same_v<SliceSpecifiers, full_extent_t> ||
+         std::is_convertible_v<SliceSpecifiers,
+                               std::tuple<IndexType, IndexType>>)&&...)) ||
+      // Preserve if leftmost SubRank-1 slices are full_extent_t and
+      // the leftmost SubRank slices are ranges
+      (((Idx < SubRank - 1 && std::is_same_v<SliceSpecifiers, full_extent_t>) ||
+        (Idx == SubRank - 1 &&
+         (std::is_same_v<SliceSpecifiers, full_extent_t> ||
+          std::is_convertible_v<SliceSpecifiers,
+                                std::tuple<IndexType, IndexType>>)) ||
```

Comment:
I think readability would be enhanced if you had something like `is_range_specifier(SliceSpecifier)` since this is repeated a few times